### PR TITLE
fix(tiles): i3s traverser for multiple viewports

### PR DESCRIPTION
**Pull request by belom88**
We got error in I3SDebug. Traversal freeze when we use 2 viewports feature.

### CODE REVIEW

**Review by ibgreen — COMMENTED** [review]
Why is this code needed for I3S but not for 3D tiles? Should it be moved into the generic tile traversal classes?

This code seems to add another layer of complexity but there are no tests.

**Comment by belom88** [comment]
> Why is this code needed for I3S but not for 3D tiles? Should it be moved into the generic tile traversal classes?

I3S has different node structure. Every node's json is requested separately. TileManager is to handle that multiple requests. It is only applicable for I3S.

File: modules/tiles/src/tileset/traversers/i3s-tile-manager.ts [hunk 1]
```diff
@@ -18,23 +18,32 @@ export default class I3STileManager {
 
   add(request, key, callback, frameState: FrameState) {
     if (!this._statusMap[key]) {
-      const {frameNumber} = frameState;
+      const {
+        frameNumber,
+        viewport: {id}
```
[the remark this answers is not in the document]
Comment (belom88) [reply modules/tiles/src/tileset/traversers/i3s-tile-manager.ts:23]: viewport come from deck.gl side. We use id from it to link a request with a viewport

File: modules/tiles/src/tileset/traversers/tileset-traverser.ts [hunk 2]
```diff
@@ -45,7 +45,7 @@ export default class TilesetTraverser {
   protected _emptyTraversalStack: ManagedArray;
   protected _frameNumber: number | null;
 
-  protected get traversalFinished(): boolean {
+  protected traversalFinished(frameState: FrameState): boolean {
```
[the remark this answers is not in the document]
Comment (belom88) [reply modules/tiles/src/tileset/traversers/tileset-traverser.ts:48]: It should be good. I designed this method to always `return true` for 3DTiles to preserve existing logic.

File: modules/tiles/src/tileset/traversers/i3s-pending-tiles-register.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export default class I3SPendingTilesRegister {
+  private frameNumberMap: Map<number, Map<number, number>> = new Map();
+
+  /**
+   * Register a new pending tile header for the particular frameNumber
+   * @param viewportId
+   * @param frameNumber
+   */
+  register(viewportId: number, frameNumber: number) {
```
[the remark this answers is not in the document]
Comment (belom88) [reply modules/tiles/src/tileset/traversers/i3s-pending-tiles-register.ts]: Done

File: modules/tiles/src/tileset/traversers/i3s-tile-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (belom88) [reply modules/tiles/src/tileset/traversers/i3s-tile-manager.ts:30]: I created `Viewport` type and added JSDoc

**Comment by belom88** [comment]
I added tests for the [i3s-pending-tiles-register.ts](https://github.com/visgl/loaders.gl/pull/2213/files/65ccb9728a0382c41c7a6721a38043e1d26953f4#diff-4b0d3b473ffe446f1cfd0175e3aa785918f3881ed3fb0b1eaa8145ae46f1f891)

**Review by ibgreen — APPROVED** [review]

**Comment by belom88** [comment]
> @belom88 I've come across a type error as the `Viewport` type added in this PR doesn't match the deck.gl type. Some properties are not present in deck at all (e.g. `cameraUp`) and others are only present on the web mercator viewport (`latitude`/ `longitude`). I think we could tighten up the type requirement on the deck side (e.g. enforce that the viewports being passed are `WebMercatorViewport` types, but I also think it would be good to have the type definition in loaders.gl actually match that of deck.gl (e.g. remove the `cameraUp` and similar fields).

Yes. It would be good. Could you create an issue?

**Pull request merged**